cookie-authentication.md

```diff
@@ -0,0 +1,145 @@
+<a name="#top">Back to Top</a>
+
+# Cookie Authentication
+
+When a user authenticates with a web browser, we need to persist that authentication
+result so that the user does not have to present their credentials on every request.
+To achieve this we will use our OAuth2 password grant flow, and store the
+tokens in cookies.
+
+**NOTE**: This is *not* a "session" implementation, there is no server-side
+state being managed for the developer.  This is pure authentication, using
+cookies as the secure storage location on the client.
+
+### Authentication Flow
+
+1. The user visits the login page and submits their username and password
+
+2. The integration uses the SDK to perform the password grant exchange, using
+    the Stormpath Application's `/oauth/token` endpoint.
+
+3. The integration collects the access token and refresh token from the SDK
+    response in the previous step.  It then responds to the browser request by
+    settings these values in cookies.
+
+4. Subsequent requests by the browser will supply these cookies which contain the
+    access and refresh tokens.  The integration should perform the following
+    logic, using the SDK's relevant JWT and OAuth authenticators:
+
+  * If the access token is valid, authenticate the request
+  * If the access token is invalid, attempt to use the refresh token to get a
+  new access token
+  * If a new access token is obtained, authenticate the request and store this
+  new token value in the access token cookie
+  * If a new access token cannot be obtained, reject the request and delete the
+    refresh token cookie
+
+### Rejecting Requests
+
+If the request cannot be authenticated, how to respond will depend on the
+accept type of the request:
+
+`Accept: text/html` - respond with a 302 redirect to the login view, with
+the requested url provided as the `?next=` parameter
+
+`Accept: application/json` - respond with `401 Unauthorized` and an empty
+body
+
+### Cookie Flags
+
+For each cookie, access token and refresh token, use these rules to determine
+the flags to set on the cookies when you are sending them to the browser:
+
+* The `expires` flag should be determined by the `ttl` value for the
+token, as defined by the Oauth Policy of the Stormpath Application.
+
+* The `HttpOnly` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.httpOnly` is false, do not set this flag
+  * Othwerwise, set this flag to true
+
+* The `path` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.path` is defined, use that value
+  * Othwerwise, use `web.basePath` if defined
+  * Finally, fallback to "/" as the value
+
+* The `domain` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.domain` is defined, use that value
```

Comment:
A `null` value would indicate 'let the web framework set the domain`.

I'm not sure if `web.domain` is required.